drivers/sensor/mma8451q/mma8451q.c

```diff
@@ -0,0 +1,191 @@
+				   enum sensor_channel chan,
+				   struct sensor_value *val)
+{
+	const struct mma8451q_config *config = dev->config->config_info;
+	struct mma8451q_data *data = dev->driver_data;
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+		mma8451q_accel_convert(val, data->x, config->range);
+		break;
+
+	case SENSOR_CHAN_ACCEL_Y:
+		mma8451q_accel_convert(val, data->y, config->range);
+		break;
+
+	case SENSOR_CHAN_ACCEL_Z:
+		mma8451q_accel_convert(val, data->z, config->range);
+		break;
+
+	case SENSOR_CHAN_ACCEL_XYZ:
+		mma8451q_accel_convert(val++, data->x, config->range);
+		mma8451q_accel_convert(val++, data->y, config->range);
+		mma8451q_accel_convert(val++, data->z, config->range);
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api mma8451q_driver_api = {
+	.sample_fetch = mma8451q_sample_fetch,
+	.channel_get = mma8451q_channel_get,
+};
+
+int mma8451q_init(struct device *dev)
+{
+	const struct mma8451q_config *config = dev->config->config_info;
+	struct mma8451q_data *data = dev->driver_data;
+
+	data->i2c = device_get_binding(config->i2c_name);
+	if (data->i2c == NULL) {
+		SYS_LOG_ERR("Failed to get pointer to %s device!",
+				config->i2c_name);
+		return -EINVAL;
+	}
+
+	u8_t whoami;
+
+	if (i2c_reg_read_byte(data->i2c, config->i2c_address,
+			      MMA8451Q_REG_WHOAMI, &whoami)) {
+		SYS_LOG_ERR("Could not get WHOAMI value");
+		return -EIO;
+	}
+
+	if (whoami != config->whoami) {
+		SYS_LOG_ERR("WHOAMI value received 0x%x, expected 0x%x",
+			    whoami, MMA8451Q_REG_WHOAMI);
+		return -EIO;
+	}
+
+	if (i2c_reg_write_byte(data->i2c, config->i2c_address,
+				   MMA8451Q_CTRL_REG1, 0) < 0) {
+		SYS_LOG_ERR("Could not set accel in config mode");
+		return -EIO;
+	}
+
+	if (i2c_reg_write_byte(data->i2c, config->i2c_address,
+				   MMA8451Q_XYZ_DATA_CFG, config->range) < 0) {
+		SYS_LOG_ERR("Could not set range");
+		return -EIO;
+	}
+
+	if (i2c_reg_write_byte(data->i2c, config->i2c_address,
+				   MMA8451Q_CTRL_REG2, 0) < 0) {
+		SYS_LOG_ERR("Could not set to normal mode");
+		return -EIO;
+	}
+
+	if (i2c_reg_write_byte(data->i2c, config->i2c_address,
+				   MMA8451Q_CTRL_REG3, 0) < 0) {
+		SYS_LOG_ERR("Could not set to low polarity, push-pull output");
+		return -EIO;
+	}
+
+	if (i2c_reg_write_byte(data->i2c, config->i2c_address,
+				   MMA8451Q_CTRL_REG4, 1) < 0) {
+		SYS_LOG_ERR("Could not enable interrupt for data ready");
+		return -EIO;
+	}
+
+	if (i2c_reg_write_byte(data->i2c, config->i2c_address,
+				   MMA8451Q_CTRL_REG1, 0x09) < 0) {
+		SYS_LOG_ERR("Could not set data rate to 800Hz");
+		return -EIO;
+	}
+
+	SYS_LOG_DBG("Init complete");
+
+	return 0;
+}
+
+static const struct mma8451q_config mma8451q_config = {
+	.i2c_name = CONFIG_MMA8451Q_I2C_NAME,
+	.i2c_address = CONFIG_MMA8451Q_I2C_ADDRESS,
+	.whoami = CONFIG_MMA8451Q_WHOAMI,
+	.start_addr = MMA8451Q_OUT_X_MSB,
+	.start_channel = MMA8451Q_CHANNEL_ACCEL_X,
+	.num_channels = MMA8451Q_NUM_ACCEL_CHANNELS,
```

Comment:
start_addr, start_channel, and num_channels are overkill for this driver, since it's just a 3-axis accelerometer. They're in the fxos8700 driver to handle accelerometer-only,  magnetometer-only, and hybrid (accel+mag) modes.

Reply:
I thought it looked very clean (that's why I adopted it) - also, there are more settings that could be used later - but ok, I can make it defines, no probs.